src/stream_io.rs

```diff
@@ -13,10 +11,9 @@ pub unsafe extern "C" fn write_stream<T: Write>(writer: *mut c_void,
                                                 n: c_int)
                                                 -> c_int {
     let bytes = slice::from_raw_parts(c as *const u8, n as usize);
-    if let Ok(_) = (*(writer as *mut T)).write_all(bytes) {
-        return 0;
-    } else {
-        return 1;
+    match (*(writer as *mut T)).write_all(bytes) {
+        Ok(_) => return 0,
+        Err(e) => return e.raw_os_error().unwrap_or(1) as c_int,
```

Comment:
I separated out the error code from the success flag, and also distinguish between system errors and otherwise.  I'm thinking this is good enough for now?